interp/src/stk_env.rs

```diff
@@ -645,6 +699,147 @@ impl<K: Eq + std::hash::Hash, V: Eq> Smoosher<K, V> {
+    /// Also panics if either is empty, or if the sets to be merged are disjoint (to be reviewed)
+    /// ```
+    /// # Examples
+    /// ## Pictorial Example
+    /// ```text
+    /// Smoosher 1: (A, B, C, F, G), Smoosher 2: (D, E, F, G), Smoosher 3: (J, I, H, F, G)       
+    /// [A]       [J]
+    ///  |        |
+    /// [B]      [I]   [D]
+    ///  |        \    |
+    /// [C]      [H]  [E]
+    ///  |       |    |
+    ///   \     |    /
+    ///    \   |   /
+    ///      [F]
+    ///       |
+    ///      [G]
+    /// *merge(Smoosher_1, [Smoosher_2, Smoosher_3])
+    /// Returns:
+    ///      [ABCDEHIJ merged and smooshed onto F]
+    ///       |
+    ///      [G]
+    /// ```
+    /// ## Code Example
+    /// good:
+    /// ```
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hello", 15);
+    /// a.set("hi!", 1);
+    /// a.new_scope();
+    /// a.set("bye", 0);
+    /// let mut b = a.fork();
+    /// let mut c = a.fork_from_tail();
+    /// a.set("hi!", 3);
+    /// b.set("bye", 2);
+    /// c.set("privet", 11);
+    /// let mut lst = Vec::new();
+    /// lst.push(b);
+    /// lst.push(c);
+    /// let d = Smoosher::merge_many(a, lst);
+    /// assert_eq!(*d.get(&"privet").unwrap(), 11);
+    /// assert_eq!(*d.get(&"hi!").unwrap(), 3);
+    /// assert_eq!(*d.get(&"bye").unwrap(), 2);
+    /// assert_eq!(*d.get(&"hello").unwrap(), 15);
+    /// ```
+    /// will panic:
+    /// ```should_panic
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hello", 15);
+    /// a.set("hi!", 1);
+    /// a.new_scope();
+    /// a.set("bye", 0);
+    /// let mut b = a.fork();
+    /// a.set("hi!", 3);
+    /// let mut c = a.fork();
+    /// b.set("bye", 2);
+    /// c.set("privet", 11);
+    /// let mut lst = Vec::new();
+    /// lst.push(b);
+    /// lst.push(c);
+    /// let d = Smoosher::merge_many(a, lst); //a and b has a different fork point
+    //from a and c
+    /// ```
+    pub fn merge_many(self, other: Vec<Self>) -> Self {
+        //initialize all needed variables
+        let mut a = self;
+        //needed to check for common fork point for all smooshers
+        let mut dp_first = 0;
+        let mut smooshed = Vec::new();
+
+        //iterate over all the smooshers and check for common fork point for all
+        //of them as well as smoosh them all to one level above the common fork.
+        for sm in other {
+            if let Some((depth_a, depth_b)) =
+                Smoosher::shared_fork_point(&a, &sm)
+            {
+                if dp_first == 0 {
+                    dp_first = depth_a;
+                    smooshed.push(sm.smoosh(depth_b - 1));
+                } else {
+                    if dp_first != depth_a {
+                        panic!(
+                            "The common fork differs for one or more smooshers"
+                        )
```

Comment:
May be worth remarking in the documentation that the shared fork point must be the _first_ shared point across all given branches (i.e., there are no pairs of branches with a common fork point which is "later" than the common fork point being used for the whole merge)

Reply:
Alternatively, if we do want to support a more robust merge-many then the code would need to identify which sets of branches must first be merged according to their shared history before the top level merge can occur, though this is definitely more complicated. It is not clear to me that this functionality is needed given the way par blocks are nested, but worth remembering that this is possible in the abstract and we should be clear about whether or not it is supported